src/fm/cp_cfm_basic_linalg.F

```diff
@@ -61,6 +66,87 @@ MODULE cp_cfm_basic_linalg
 
 CONTAINS
 
+! **************************************************************************************************
+!> \brief Computes the determinant (with a correct sign even in parallel environment!) of a complex square matrix
+!> \param matrix_a ...
+!> \param det_a ...
+!> \author A. Sinyavskiy (andrey.sinyavskiy@chem.uzh.ch)
+! **************************************************************************************************
+   SUBROUTINE cp_cfm_det(matrix_a, det_a)
+
+      TYPE(cp_cfm_type), INTENT(IN)            :: matrix_a
+      COMPLEX(KIND=dp), INTENT(OUT)            :: det_a
+      COMPLEX(KIND=dp)                         :: determinant
+      TYPE(cp_cfm_type)                        :: matrix_lu
+      COMPLEX(KIND=dp), DIMENSION(:, :), POINTER  :: a
+      INTEGER                                  :: n, i, info, P
+      INTEGER, ALLOCATABLE, DIMENSION(:)       :: ipivot
+      COMPLEX(KIND=dp), DIMENSION(:), POINTER  :: diag
+
+#if defined(__SCALAPACK)
+      INTEGER, EXTERNAL                        :: indxl2g
+      INTEGER                                  :: myprow, nprow, npcol, nrow_local, nrow_block, irow_local, &
+                                                  mypcol, ncol_local, ncol_block, icol_local, j
+      INTEGER, DIMENSION(9)                    :: desca
+#endif
+
+      CALL cp_cfm_create(matrix=matrix_lu, &
+                         matrix_struct=matrix_a%matrix_struct, &
+                         name="A_lu"//TRIM(ADJUSTL(cp_to_string(1)))//"MATRIX")
+      CALL cp_cfm_to_cfm(matrix_a, matrix_lu)
+
+      a => matrix_lu%local_data
+      n = matrix_lu%matrix_struct%nrow_global
+      ALLOCATE (ipivot(n))
+      ipivot(:) = 0
+      P = 0
+      ALLOCATE (diag(n))
+      diag(:) = 0.0_dp
+#if defined(__SCALAPACK)
+      ! Use LU decomposition
+      desca(:) = matrix_lu%matrix_struct%descriptor(:)
+      CALL pzgetrf(n, n, a(1, 1), 1, 1, desca, ipivot, info)
+      myprow = matrix_lu%matrix_struct%context%mepos(1)
+      mypcol = matrix_lu%matrix_struct%context%mepos(2)
+      nprow = matrix_lu%matrix_struct%context%num_pe(1)
+      npcol = matrix_lu%matrix_struct%context%num_pe(2)
+      nrow_local = matrix_lu%matrix_struct%nrow_locals(myprow)
+      nrow_block = matrix_lu%matrix_struct%nrow_block
+      ncol_block = matrix_lu%matrix_struct%ncol_block
+      ncol_local = matrix_lu%matrix_struct%ncol_locals(mypcol)
+
+      DO irow_local = 1, nrow_local
+         i = indxl2g(irow_local, nrow_block, myprow, matrix_lu%matrix_struct%first_p_pos(1), nprow)
+         DO icol_local = 1, ncol_local
+            j = indxl2g(icol_local, ncol_block, mypcol, matrix_lu%matrix_struct%first_p_pos(2), npcol)
+            IF (i == j) diag(i) = matrix_lu%local_data(irow_local, icol_local)
+         END DO
+      END DO
+      CALL matrix_lu%matrix_struct%para_env%sum(diag)
```

Comment:
I think that there is no need to exchange the diagonal elements. Just multiply them in some container (let's say `determinant`) and exchange it using something like `para_env%prod(determinant)`.